tools/cli.py

```diff
@@ -0,0 +1,195 @@
+        self._args = []
+
+    @property
+    def state(self):
+        return (
+            self._channel, self._tty, self.prompt, self._quiet,
+            self._term, self._level, self._commands)
+
+    @state.setter
+    def state(self, state):
+        (self._channel, self._tty, self.prompt, self._quiet, self._term,
+         self._level, self._commands) = state
+
+    def cmd(self, command_set, delete=False):
+        """ update command list """
+        if delete:
+            self._commands = {
+                "pause": (
+                    self._Pause,
+                    "<text.S>",
+                    "Print string and wait for Enter key"),
+                "echo": (
+                    self._Echo,
+                    "<text.S>",
+                    "Print string"),
+                "quit": (
+                    self._Quit,
+                    "",
+                    "Quit"),
+                "help": (
+                    self._Help,
+                    "",
+                    "Provide help"),
+                "@": (
+                    self._At,
+                    "<file.F> [quiet]",
+                    "Read commands from file"),
+                "?": (
+                    self._Query,
+                    "",
+                    "List commands"),
+            }
+        self._commands.update(command_set)
+
+    def commands_starting_with(self, string):
+        for c in self._commands:
+            if c.startswith(string):
+                yield c
+
+    def __prompt(self):
+        if self._tty and self._term is None:
+            print(self.prompt, end="", flush=True)
+
+    def run(self):
+        """ execute a CLI """
+        self.__prompt()
+        for line in self._channel:
+            if not self._tty and not self._quiet:
+                print(self.prompt + line)
+            print("")
+            line = line.strip()
+            if not line or line.startswith("#"):
+                self.__prompt()
+                continue
+            self._args = line.split()
+            cmd = self._args.pop(0)
+            if cmd in self._commands:
+                try:
+                    if self._commands[cmd][0](self):
+                        break
+                except Exception as e:  # pylint: disable=broad-except
+                    print("error: {}".format(e))
+            else:
+                print("bad command \"{}\"".format(cmd))
+            self.__prompt()
+
+    def __write(self, text):
+        print(text, flush=self._tty)
+
+    @property
+    def count(self):
+        return len(self._args)
+
+    @property
+    def args(self):
+        for a in self._args:
+            yield a
+
+    def arg(self, n):
+        return self._args[n]
+
+    def arg_i(self, n):
+        return int(self._args[n], base=0)
+
+    def arg_x(self, n):
+        return int(self._args[n], base=16)
+
+    # Default commands of a CLI object; always present!
+
+    def _Pause(self):
```

Comment:
Why underscore Captial name?  Same elsewhere

Reply:
Capital to match ybug, `_` because it's internal to the class (it was added later).

Reply:
Changed `_Pause` to `_cmd_pause` (and analogous other methods).